### ISSUE TITLE
Loading and using a dependency in a package's template?

### ISSUE THREAD

**Issue by Aerlinger** [cocos-creator/editor-framework#398]
Where is the proper location to load any dependencies needed in a package's renderer?

For instance, in the following code editor example, `CodeMirror` is a dependency used in the package's template. 

The `CodeMirror` object must be defined before it can be used, however, the following fails:

Example:
```javascript
Editor.Panel.extend({
  template: `
        <!-- Where/how should this dependency be loaded? -->
        <script src="${Editor.url('packages://code-editor/node_modules/codemirror/lib/codemirror.js')}"></script>

        <h2>Code Editor</h2>
        <ui-text-area id="code_editor"></ui-text-area>
    `,

  $: {
    editor: "#editor",
    code_editor: "#code_editor"
  },

  listeners: {
    open() {
      Editor.Panel.open('code-editor')
    }
  },

  ready () {
    // CodeMirror Not defined!?
    let editor = CodeMirror.fromTextArea(this.$code_editor, {
      mode:        "python",
      lineNumbers: true
    });

    editor.reflow()
  },

  run (argv) {
    console.log(`Pkg run with args: ${argv}`);
  }
});
```

The above yields an error: `ReferenceError: CodeMirror is not defined`. 

What's the proper approach for including this sort of dependency?

**Comment by Aerlinger** [comment]
Also, I realize one solution is to require the library directly from the `ready` callback:

```javascript
ready () {
    let CodeMirror = require("codemirror")

    // Error goes away, but still missing CSS imports
    let editor = CodeMirror.fromTextArea(this.$code_editor, {
      mode:        "python",
      lineNumbers: true
    });
  },
```

However, I still have the problem that the CSS doesn't load properly.

**Comment by Aerlinger** [comment]
So, I did edit the template with the following, which does work:

```javascript
template: `
        <style> @import url('${Editor.url('packages://code-editor/node_modules/codemirror/lib/codemirror.css')}') </style>

        <textarea id="code_editor"></textarea>
    `,
```

But this feels very hacky, and not future-proof. Any thoughts on a better approach.

**Comment by jwu** [comment]
you can use the keyword dependencies.

For example:

```javascript
Editor.Panel.extend({
  dependencies: [
    'packages://code-editor/node_modules/codemirror/lib/codemirror.js'
  ]
});
```

**Comment by jwu** [comment]
> Also, I realize one solution is to require the library directly from the ready callback:

Em, how can you use `require('codemirror')` without can't find the module problem? Since codemirror should be `npm install` in your package, so how can Editor-Framework correctly find the path of code-mirror ?

**Comment by jwu** [comment]
For the `<style>`, I think I need to add ability to `style` keyword, so that it can accept arrays with file paths in some how.

**Comment by Aerlinger** [comment]
> Em, amazing... how can you use require('codemirror') without "can't find the module" problem? 

I assumed this was because of Electron, not Editor-Framwork:
https://discuss.atom.io/t/using-require-in-electron-renderer-process-code/21911

**Comment by Aerlinger** [comment]
> The first one will cache in Editor.UI and use async loading

Hm, would this ever be desirable. Wouldn't this mean `ready()` could be called before the  module actually loaded?

**Comment by jwu** [comment]
`ready()` will be call after all dependencies loaded.

**Comment by jwu** [comment]
> So, I did edit the template with the following, which does work: .... But this feels very hacky, and not future-proof. Any thoughts on a better approach.

I don't think you are hack the code. Actually editor-framework did couple of similar things in ui element styles. But here you can make it simple and NOT that hacky:

```javascript
template: `
        <style> 
          @import url(packages://code-editor/node_modules/codemirror/lib/codemirror.css')                    
        </style>
        <textarea id="code_editor"></textarea>
    `,
```

In css, html, all custom protocol are welcome, and they did run into the custom protocol registry handler.

**Comment by jwu** [comment]
> I assumed this was because of Electron, not Editor-Framework:
https://discuss.atom.io/t/using-require-in-electron-renderer-process-code/21911

My curious comes from your code: `packages://code-editor/node_modules/...` which means you `npm install` code-mirror in the code-editor packages. Electron should not realise that path, so how can it find it? interesting. 😕
